elements/lisk-chain/src/diff.ts

```diff
@@ -130,15 +130,104 @@ const diffAlgo = (initial: Buffer, final: Buffer): HistoryType[] => {
 	return [];
 };
 
+/**
+ * This function returns the length of common prefix between two buffers
+ */
+const diffCommonPrefix = (buffer1: Buffer, buffer2: Buffer): number => {
+	// Quick check for common null cases.
+	if (buffer1[0] !== buffer2[0]) {
+		return 0;
+	}
+	// Binary search.
+	// Performance analysis: http://neil.fraser.name/news/2007/10/09/
+	let pointerMin = 0;
+	let pointerMax = Math.min(buffer1.length, buffer2.length);
+	let pointerMid = pointerMax;
+	let pointerstart = 0;
+
+	while (pointerMin < pointerMid) {
+		if (
+			buffer1
+				.slice(pointerstart, pointerMid)
+				.equals(buffer2.slice(pointerstart, pointerMid))
+		) {
+			pointerMin = pointerMid;
+			pointerstart = pointerMin;
+		} else {
+			pointerMax = pointerMid;
+		}
+		pointerMid = Math.floor((pointerMax - pointerMin) / 2 + pointerMin);
+	}
+
+	return pointerMid;
+};
+
+/**
+ * This function returns the length of common suffix between two buffers
+ */
+const diffCommonSuffix = (buffer1: Buffer, buffer2: Buffer): number => {
+	// Quick check for common null cases.
+	if (buffer1[buffer1.length - 1] !== buffer2[buffer2.length - 1]) {
+		return 0;
+	}
+	// Binary search.
+	// Performance analysis: http://neil.fraser.name/news/2007/10/09/
+	let pointerMin = 0;
+	let pointerMax = Math.min(buffer1.length, buffer2.length);
+	let pointerMid = pointerMax;
+	let pointerEnd = 0;
+
+	while (pointerMin < pointerMid) {
+		if (
+			buffer1
+				.slice(buffer1.length - pointerMid, buffer1.length - pointerEnd)
+				.equals(
+					buffer2.slice(
+						buffer2.length - pointerMid,
+						buffer2.length - pointerEnd,
+					),
+				)
+		) {
+			pointerMin = pointerMid;
+			pointerEnd = pointerMin;
+		} else {
+			pointerMax = pointerMid;
+		}
+		pointerMid = Math.floor((pointerMax - pointerMin) / 2 + pointerMin);
+	}
+
+	return pointerMid;
+};
+
 /**
  * This function reduces the diff generated by the diff algorithm based on unchanged bytes
  */
 export const calculateDiff = (
 	initial: Buffer,
 	final: Buffer,
 ): HistoryType[] => {
-	const longDiff = diffAlgo(initial, final);
-	const reducedDiff = [];
+	// When both the buffers are equal then return all '=' history
+	if (initial.equals(final)) {
+		return [['=', initial.length]];
+	}
+	const commonPrefix = diffCommonPrefix(initial, final);
+	const commonSuffix = diffCommonSuffix(initial, final);
+
+	const strippedPrefixInitial = initial.slice(commonPrefix, initial.length);
+	const strippedPrefixFinal = final.slice(commonPrefix, final.length);
+
+	const strippedInitial = strippedPrefixInitial.slice(
+		0,
+		strippedPrefixInitial.length - commonSuffix,
+	);
+	const strippedFinal = strippedPrefixFinal.slice(
+		0,
+		strippedPrefixFinal.length - commonSuffix,
+	);
```

Comment:
Here I think it might be better to:

const commonPrefix = diffCommonPrefix(initial, final);
const strippedPrefixInitial = initial.slice(commonPrefix, initial.length);
const strippedPrefixFinal = final.slice(commonPrefix, final.length);

const commonSuffix = diffCommonSuffix(strippedPrefixInitial, strippedPrefixFinal);
const strippedInitial = strippedPrefixInitial.slice(
	0,
	strippedPrefixInitial.length - commonSuffix,
);
const strippedFinal = strippedPrefixFinal.slice(
	0,
	strippedPrefixFinal.length - commonSuffix,
);

This way, computing the common suffix is done on the shorter buffers (always good) and I think we avoid some corner cases where the prefix and suffix overlap in the shorter buffer.